rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,309 @@
+    },
+})
+if err != nil {
+    return trace.Wrap(err)
+}
+defer watch.Close()
+
+// Loop forever watching for client tools version changes.
+for {
+    select {
+    case e := <-watch.Events():
+        if e.Resource == nil {
+            continue
+        }
+        resource, ok := e.Resource.(types.ClusterMaintenanceConfig)
+        if !ok {
+            continue
+        }
+
+        // Process the version change by notifying downstream tooling (like
+        // Jamf).
+        if err := handle(resource); err != nil {
+            log.Printf("Failed to handle version change: %v", err)
+        }
+    case <-watch.Done():
+        fmt.Println("The watcher job is finished.")
+        return nil
+    }
+}
+```
+
+##### Cluster configuration
+
+> [!NOTE]
+> This resource will not be configurable on Teleport Cloud. It will be managed
+> by the Teleport Cloud team.
+
+The `types.ClusterMaintenanceConfig` resource will be updated to include a
+`tools_version` field. This field can be updated directly from the resource or
+using `tctl`.
+
+```
+kind: cluster_maintenance_config
+spec:
+  # tools_version is the version of client tools the proxy will advertise. Can
+  # be auto (match the version of the proxy), off (don't advertise any version),
+  # or an exact semver formatted version.
+  tools_version: auto | off | X.Y.Z
+
+  [...]
+```
+
+```
+$ tctl autoupdate update --set-tools-version=1.0.1
+Automatic updates configuration has been updated.
+```
+
+Three modes will be supported: `auto`, `X.Y.Z` (exact version), and `off`.
+
+* `auto`: Cluster will advertise the version of the proxy as the version of
+  client tools.
+* `X.Y.Z`: Cluster will advertise a specific version of client tools.
+* `off`: Cluster will not advertise any version, clients will not try to
+  update.
+
+All new self-hosted clusters will be opted into `auto`. This way, by default
+even self-hosted customers will have automatic updates available.
+
+Existing clusters will be opted into `off` to not introduce unexpected
+behavior.
+
+The `tools_version` field will then be available from the unauthenticated
+endpoint `/v1/webapi/ping`.
+
+```
+curl https://proxy.example.com/v1/webapi/ping | jq .
+{
+    "tools_version": "X.Y.Z",
+
+    [...]
+}
+```
+
+TODO(russjones): Teleport Connect support.
+
+### Costs
+
+Some additional costs will be incurred as Teleport downloads will increase in
+frequency.
+
+### Out of scope
+
+How Cloud will push changes to `cluster_maintenance_config` is out of scope
+for this RFD and will be handled by a separate Cloud specific RFD.
+
+### Security
+
+The initial version of automatic updates will rely on TLS to establish
+connection authenticity to the Teleport download server. The authenticity of
+assets served from the download server is out of scope for this RFD. Cluster
+administrators concerned with the authenticity of assets served from the
+download server can use self-managed updates with system package managers which
+are signed.
```

Comment:
Do we want to at least validate the hash/signature on download?
- we will very likely need to expose a way for agent AU to get the public key from Teleport so it can validate the downloaded content. Local tooling installer could benefit from this
- We download/write corrupted binaries, especially as we increase the number of downloads. Checking hash after download would prevent this
- TLS is not very strong in corporate environments where each company trusts its own internal CA on the machines. Someone compromising the internal CA would be able to have code execution on any developer laptop.

Reply:
Existing discussion with Jent: https://github.com/gravitational/teleport/pull/30317#discussion_r1297653526

Reply:
I don't have a problem with checking the hash to validate the binary has not been corrupted.

I don't think we should develop a new signing scheme, because by the time we end up mitigating all the edge cases, we end up at TUF anyway. Might as well implement TUF as part of Phase 2.